### PR TITLE
3803 lncc loss fix and warp to cache regular grid

### LINKED ISSUE #3803: LNCC loss instability for large kernel size
### Discussed in https://github.com/Project-MONAI/MONAI/discussions/3463

<div type='discussions-op-text'>

<sup>Originally posted by **ebrahimebrahim** December  9, 2021</sup>
I experienced some numerically unstable training while using LNCC loss to train a registration network. Exploding gradients, clipping them not helping, training loss increasing even with a low learning rate, etc.

This is in the context of synthetic data, where the images contained large homogeneous blobs, and so the variances used to compute LNCC were probably quite low.

I see that [small variance has been a problem in the past](https://github.com/Project-MONAI/MONAI/issues/1868) and that [it has been addressed](https://github.com/Project-MONAI/MONAI/pull/1915). However, I seem to still have the issue when I set the `kernel_size` argument for LNCC loss to be large. Setting it back to a small value like `3` fixes the problem completely. Setting it to a large value (say, half the image size) leads to problems.

I think that giving a large `kernel_size` to LNCC is a natural thing to do, because NCC (normalized cross correlation, without the "local") is the case of having `kernel_size` equal image size, if I'm understanding correctly.

_Is there any reason to expect that large kernel size would worsen the instability problem?_ From [the code](https://github.com/Project-MONAI/MONAI/blob/dev/monai/losses/image_dissimilarity.py#L50), it's not clear to me why the LNCC kernel size would matter. Thoughts, anyone?</div>

### PR DESCRIPTION
**Pull request by wyli**
Fixes #3803

### Description

- fixes lncc loss
- adds a cache of the reference grid for `warp`

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by ebrahimebrahim** [comment]
Really happy to see this!

Setting the default `smooth_nr` to `0` agrees with [some earlier observations](https://github.com/Project-MONAI/MONAI/discussions/3463#discussioncomment-2899507) on this issue

**Review by KumoLiu — APPROVED** [review]
